src/io/bytes.rs

```diff
@@ -0,0 +1,69 @@
+//! Implementations of bytes::BufMut and bytes::Buf for Ring
+
+use std::io::BufRead;
+
+use bytes::{buf::UninitSlice, Buf, BufMut};
+
+use super::{Ring, SeqRead, SeqWrite};
+
+impl Buf for Ring {
+    fn remaining(&self) -> usize {
+        self.read_len()
+    }
+
+    fn chunk(&self) -> &[u8] {
+        self.as_read_slice(usize::MAX)
+    }
+
+    fn advance(&mut self, cnt: usize) {
+        self.consume(cnt)
+    }
+}
+
+unsafe impl BufMut for Ring {
+    fn remaining_mut(&self) -> usize {
+        self.write_len()
+    }
+
+    unsafe fn advance_mut(&mut self, cnt: usize) {
+        self.feed(cnt)
+    }
+
+    fn chunk_mut(&mut self) -> &mut UninitSlice {
+        UninitSlice::new(self.as_write_slice(usize::MAX))
```

Comment:
this is probably the weirdest part of the BufMut API, but as documented here `::new` is used for initialized memory. https://docs.rs/bytes/latest/bytes/buf/struct.UninitSlice.html#method.new